File: scripts/oop-practice-sources/java/figure.java
// 1. Создаем чертеж (класс) для любой фигуры
class Figure {
    String name;
    String color;

    Figure(String name, String color) {
        this.name = name;
        this.color = color;
    }

    void describe() {
        System.out.println("Я " + color + " " + name);
    }
}

public class Main {
    public static void main(String[] args) {
        // 2. Создаем конкретные фигуры (объекты)
        Figure circle = new Figure("Круг", "Красный");
        Figure square = new Figure("Квадрат", "Синий");

        // 3. Используем их
        circle.describe();   // Я Красный Круг
        square.describe();   // Я Синий Квадрат
    }
}

File: scripts/oop-practice-sources/rust/figure.rs
struct Figure {
    name: String,
    color: String,
}

impl Figure {
    fn describe(&self) {
        println!("Фигура «{}», цвет: {}", self.name, self.color);
    }
}

struct Circle {
    figure: Figure,
}

impl Circle {
    fn new(color: &str) -> Self {
        Self {
            figure: Figure {
                name: "Круг".to_string(),
                color: color.to_string(),
            },
        }
    }

    fn describe(&self) {
        self.figure.describe();
    }
}

struct Square {
    figure: Figure,
}

impl Square {
    fn new(color: &str) -> Self {
        Self {
            figure: Figure {
                name: "Квадрат".to_string(),
                color: color.to_string(),
            },
        }
    }

    fn describe(&self) {
        self.figure.describe();
    }
}

fn main() {
    let circle = Circle::new("красный");
    let square = Square::new("синий");
    circle.describe();
    square.describe();
}
